Adding rules for TEA, XTEA, XXTEA and RIPEMD hashing

raymondlleong opened the pull request:
Adding rules for XTEA, XXTEA, RIPEMD-128, 160, 256, 320
Adding to current rules for TEA encrypt & decrypt

Relates to issue #676

williballenthin commented on nursery/hash-data-using-ripemd128.yml:
```diff
@@ -0,0 +1,26 @@
+          - number: 0x76543210 = RIPEMD-256 and RIPEMD-320 constant
+          - number: 0xFEDCBA98 = RIPEMD-256 and RIPEMD-320 constant
+          - number: 0xC3D2E1F0 = RIPEMD-160 and RIPEMD-320 constant
+          - number: 0x3C2D1E0F = RIPEMD-320 constant
```
is there a way to exclude MD5 here?

williballenthin replied:
i don't totally understand all the cases here and if you've covered them, but i trust you to have done that work :-)

raymondlleong replied:
Good point and added an additional function scope for an array constant not present in MD5. Would have liked to keep it a function scoped rule but the magic constants are sometimes in a function that solely moves their values into a structure with nothing else to signature off:
<img width="481" alt="Screen Shot 2023-02-14 at 11 47 18 PM" src="https://user-images.githubusercontent.com/77110701/219266655-a8c3c377-a0c9-466e-8dcb-c3eba7a87538.png">

raymondlleong replied:
The RIPEMD160 rule is having a similar issue with SHA1. The magic constants listed in https://rst.im/magic/RIPEMD160 are also present in SHA1, including the K constants even though they arent listed in the corresponding SHA1 page. The routines have a lot of similarities and more research is needed to write a rule that better deconflicts the two. I think keeping the RIPEMD160 rule out of the repo may be better for now to avoid false positives. 

Of note, the logic in the SHA1 rule below will also hit on RIPEMD initialization but with SHA1 being much more common than the other two may be less of an issue:
```
- or:
      - and:
        - description: Magic initialization constants used in SHA1
        - number: 0x67452301 = A, also used by MD5
        - number: 0xEFCDAB89 = B, also used by MD5
        - number: 0x98BADCFE = C, also used by MD5
        - number: 0x10325476 = D, also used by MD5
        - number: 0xC3D2E1F0 = specific to SHA1, not MD4 nor MD5
```

williballenthin replied:
awesome, thanks for the explanation. it seems a little tricky to get perfectly right, but i tend to agree with you that we should focus on the mostly likely scenarios and being helpful to most analysts. maybe disabling one of the rules and adding extra documentation the other is best?

raymondlleong replied:
Thanks for the suggestion. Updated the documentation in the MD5 and SHA1 rules to highlight the shared constants with the RIPEMD routines and removed the RIPEMD160 rule that would be prone to false positives with SHA1 (didnt see a way to disable it).

williballenthin commented on nursery/decrypt-data-using-tea.yml:
```diff
@@ -4,14 +4,43 @@ rule:
       - Defense Evasion::Obfuscated Files or Information::Encryption-Standard Algorithm [E1027.m05]
     references:
       - https://en.wikipedia.org/wiki/Tiny_Encryption_Algorithm
+    examples:
+      - 5B99FA01C72CEBC53A76CC72E9581189:0x180001027
+      - 7FD2E2E3C88675D877190ABAA3002B55:0x010004EC0
+      - 775B26D6ACCB5A4AE868983F8F5006A3:0x010018890
```
now that we have examples, let's move this rule out of the nursery.

williballenthin replied:
in fact, all rules in this PR with examples we can move out of the nursery, awesome!

williballenthin commented on nursery/decrypt-data-using-tea.yml:
```diff
@@ -4,14 +4,43 @@ rule:
     - and:
-      - number: 0x9E3779B9 = delta
-      - number: 0xC6EF3720 = sum
+      - basic block:
+        - and:
+          - instruction:
+            - mnemonic: shr
+            - operand[1].number: 0x5
+          - instruction:
+            - mnemonic: shl
+            - operand[1].number: 0x4
+          - count(mnemonic(shr)): 0x2
+          - characteristic: nzxor
+      - or:
+        - operand[1].number: 0x9E3779B9 = key schedule constant
+        - operand[1].number: 0x61C88647 = key schedule constant two's complement
+      - operand[1].number: 0xC6EF3720 = sum
```
neat logic!

williballenthin commented on nursery/hash-data-using-ripemd256.yml:
```diff
@@ -0,0 +1,27 @@
+    authors:
+      - raymond.leong@mandiant.com
+    scope: function
+    references:
+      - https://en.wikipedia.org/wiki/RIPEMD-256
+  features:
```
if no examples, please keep in the nursery

williballenthin reviewed (changes requested):
this is great! thank you for taking the time to write these rules and find so many examples.

would you please remove the `.DS_Store` files? also, it looks like some file permissions changed. we don't really use those, so i'm not sure if it matters? not a problem if its not obvious how to fix that.

with regards to the testfiles PR, as long as we have a good example for each rule, i think there are diminishing returns for adding more, especially if the logic is basically the same. so for that large file, if the logic is well covered by the other example, would you please remove it?

thanks!

raymondlleong commented:
Thank you for the review! A follow up commit addresses these comments above:
- rules with examples are moved out of nursery 
- unneeded examples removed from rules and their example files removed from the PR for capa-testfiles
- DS_store files deleted

mr-tz commented on data-manipulation/hashing/ripemd160/hash-data-using-ripemd160.yml:
```diff
@@ -0,0 +1,27 @@
+      - or:
+        - and:
+          - description: Magic initialization constants used in RIPEMD-160
+          - number: 0x67452301 = also used in MD5, SHA1, RIPEMD-128, RIPEMD-256, and RIPEMD-320
```
I like that these inline descriptions can point analysts to the alternative algorithms!

mr-tz reviewed (approved):
LGTM, thanks!

mr-tz commented:
If @williballenthin doesn't object in the next few days, I'll merge this. Thanks, everyone!

williballenthin commented:
send it!

mr-tz commented:
Thanks, Willi and Ray!
